File: go-api/internal/infrastructure/http/middleware/jwt_auth.go
package middleware

import (
	"strings"

	"github.com/gofiber/fiber/v2"

	"matrix-api-challenge/go-api/internal/infrastructure/auth"
	"matrix-api-challenge/go-api/internal/infrastructure/http/dto"
)

const authTokenLocalKey = "authToken"

func JWTAuth(tokenService *auth.TokenService) fiber.Handler {
	return func(ctx *fiber.Ctx) error {
		authorizationHeader := ctx.Get("Authorization")
		if !strings.HasPrefix(authorizationHeader, "Bearer ") {
			return respondUnauthorized(ctx, "missing or invalid authorization header")
		}

		accessToken := strings.TrimSpace(strings.TrimPrefix(authorizationHeader, "Bearer "))
		if accessToken == "" {
			return respondUnauthorized(ctx, "missing or invalid authorization header")
		}

		if err := tokenService.ValidateAccessToken(accessToken); err != nil {
			return respondUnauthorized(ctx, "invalid or expired token")
		}

		ctx.Locals(authTokenLocalKey, accessToken)

		return ctx.Next()
	}
}

func AuthTokenFromRequest(ctx *fiber.Ctx) string {
	token, _ := ctx.Locals(authTokenLocalKey).(string)
	return token
}

func respondUnauthorized(ctx *fiber.Ctx, message string) error {
	return ctx.Status(fiber.StatusUnauthorized).JSON(dto.ErrorResponse{
		Error: message,
		Code:  "UNAUTHORIZED",
	})
}

File: node-api/src/infrastructure/http/middleware/jwtAuth.js
export function createJwtAuthMiddleware(jwtTokenService) {
  return function jwtAuth(request, response, next) {
    const authorizationHeader = request.headers.authorization ?? '';

    if (!authorizationHeader.startsWith('Bearer ')) {
      response.status(401).json({
        error: 'missing or invalid authorization header',
        code: 'UNAUTHORIZED',
      });
      return;
    }

    const accessToken = authorizationHeader.slice('Bearer '.length).trim();

    if (!accessToken || !jwtTokenService.validateAccessToken(accessToken)) {
      response.status(401).json({
        error: 'invalid or expired token',
        code: 'UNAUTHORIZED',
      });
      return;
    }

    next();
  };
}
